Bug 2021324: Change the order of Project creation and deletion

sadasu opened the pull request:
Current ordering of Project creation and teardown is causing some
timing issues resulting in the namespace to be deleted while it is
still in use by a test.

sadasu commented:
/retest

derekhiggins commented on test/extended/util/client.go:
```diff
@@ -137,9 +138,8 @@ func NewCLIWithoutNamespace(project string) *CLI {
-	g.AfterEach(cli.TeardownProject)
-	g.AfterEach(cli.kubeFramework.AfterEach)
 	g.BeforeEach(cli.kubeFramework.BeforeEach)
+	g.AfterEach(cli.kubeFramework.AfterEach)
```
Looking at the ginkgo docs(not very familiar with how this code works), It looks to me like this moves where the various blocks are being registered but doesn't actually change their order, do you instead want JustAfterEach for this line so that it is called before the TeardownProject, if that what your hoping to achieve ?

sadasu replied:
Yes, that was the intent. Let me try that. Also, seems incorrect to have the BeforeEach() and AfterEach() for a given resource to be spread around in different methods like that.

sadasu commented:
/retest

sadasu commented:
@stbenjam could you PTAL?

Fix is basically trying to make sure that the project teardown and creation happen in that order between It() calls within this (https://github.com/openshift/origin/blob/master/test/extended/cli/explain.go#L385) Describe() block.

sadasu commented:
/retest

dmage commented on test/extended/util/client.go:
```diff
@@ -137,9 +138,8 @@ func NewCLIWithoutNamespace(project string) *CLI {
-	g.AfterEach(cli.TeardownProject)
-	g.AfterEach(cli.kubeFramework.AfterEach)
 	g.BeforeEach(cli.kubeFramework.BeforeEach)
+	g.JustAfterEach(cli.kubeFramework.AfterEach)
```
kubeFramework.AfterEach deletes namespaces, so it cannot be `JustAfterEach`, it can be only `AfterEach`.

`cli.TeardownProject` works fine even if when it doesn't have a namespace. It deletes configs that are generated by ChangeUser.

dmage commented:
@sadasu can you provide an example of the failure that you want to fix?

sadasu commented:
> @sadasu can you provide an example of the failure that you want to fix?

e2e-metal-ipi-ovn-ipv6 CIs jobs frequently fail on the test:
openshift-tests.[sig-cli] oc explain should contain proper spec+status for CRDs [Suite:openshift/conformance/parallel]

On further investigation, we found that each time the failure occurs, it is for a diff CRD.
For example, looking the logs for https://prow.ci.openshift.org/view/gs/origin-ci-test/logs/periodic-ci-openshift-release-master-nightly-4.8-e2e-metal-ipi-ovn-ipv6/1453850449357574144, we see the following failure:

: [sig-cli] oc explain should contain proper spec+status for CRDs [Suite:openshift/conformance/parallel] expand_less | 6s
-- | --
fail [github.com/openshift/origin/test/extended/cli/explain.go:450]: Unexpected error:     <*errors.errorString \| 0xc002392eb0>: {         s: "failed to explain [\"alertmanagerconfigs\" \"--api-version=monitoring.coreos.com/v1alpha1\"]: exit status 1",     }     failed to explain ["alertmanagerconfigs" "--api-version=monitoring.coreos.com/v1alpha1"]: exit status 1 occurred

Looking into this error further, we see:
[It] should contain proper spec+status for CRDs [Suite:openshift/conformance/parallel]
  github.com/openshift/origin/test/extended/cli/explain.go:442
Oct 28 23:41:15.350: INFO: Checking monitoring.coreos.com/v1alpha1, Resource=alertmanagerconfigs...
Oct 28 23:41:15.350: INFO: Running 'oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1'
Oct 28 23:41:16.214: INFO: Error running /usr/bin/oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1:
StdOut>
error: couldn't find resource for "monitoring.coreos.com/v1alpha1, Kind=AlertmanagerConfig"
StdErr>
error: couldn't find resource for "monitoring.coreos.com/v1alpha1, Kind=AlertmanagerConfig"

[AfterEach] [sig-cli] oc explain
  github.com/openshift/origin/test/extended/util/client.go:140
STEP: Collecting events from namespace "e2e-test-oc-explain-8d5ph".
STEP: Found 1 events.
Oct 28 23:41:16.572: INFO: At 2021-10-28 23:41:13 +0000 UTC - event for e2e-test-oc-explain-8d5ph: {namespace-security-allocation-controller } CreatedSCCRanges: created SCC ranges
Oct 28 23:41:16.661: INFO: POD  NODE  PHASE  GRACE  CONDITIONS
Oct 28 23:41:16.661: INFO: 
Oct 28 23:41:16.920: INFO: skipping dumping cluster info - cluster too large
Oct 28 23:41:17.284: INFO: Deleted {user.openshift.io/v1, Resource=users  e2e-test-oc-explain-8d5ph-user}, err: <nil>
Oct 28 23:41:17.385: INFO: Deleted {oauth.openshift.io/v1, Resource=oauthclients  e2e-client-e2e-test-oc-explain-8d5ph}, err: <nil>
Oct 28 23:41:17.486: INFO: Deleted {oauth.openshift.io/v1, Resource=oauthaccesstokens  sha256~v995EXhoATI1rWGqF-JZFczynKhx2EEnnXNtVVaqke0}, err: <nil>
[AfterEach] [sig-cli] oc explain
  github.com/openshift/origin/test/extended/util/client.go:141
STEP: Destroying namespace "e2e-test-oc-explain-8d5ph" for this suite.
fail [github.com/openshift/origin/test/extended/cli/explain.go:450]: Unexpected error:
    <*errors.errorString | 0xc002392eb0>: {
        s: "failed to explain [\"alertmanagerconfigs\" \"--api-version=monitoring.coreos.com/v1alpha1\"]: exit status 1",
    }
    failed to explain ["alertmanagerconfigs" "--api-version=monitoring.coreos.com/v1alpha1"]: exit status 1
occurred

Looking at these logs, it appears that there is a timing issue regarding when the namespace created via the Project creation is destroyed and when the "oc explain" is run on that resource.

This is a very frequently occurring failure on the metal platform CIs although the CRDs that fail are not specific to the metal platform. We have not looked at the frequency of occurrence of these failures on other Platform CIs.

This current fix is meant to spark discussion around what the correct fix needs to be. So, @dmage thanks for taking a look at that.
I have added some of this information to the BZ too.

sadasu commented:
/retest

sadasu commented:
/retest

dmage commented:
@sadasu
1. As you are trying to fix a bug in master, please find a similar failure on 4.10.
2. `oc explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1` doesn't rely on namespaces, it works only with cluster-scoped endpoints, so project creation/deletion shouldn't affect it.

dmage commented:
```
$ oc --namespace=something-that-does-not-exist explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1
KIND:     AlertmanagerConfig
VERSION:  monitoring.coreos.com/v1alpha1

DESCRIPTION:
...
```

sadasu commented:
@dmage , this failure does not occur during 4.10 metal-ipi tests. They were occuring during 4.8 CIs pretty frequestly. And now, after https://github.com/openshift/release/pull/23131 has merged, we haven't seen this failure. This is meant to be a temporary fix but points to API server unavailability as a possible reason the "oc explain" tests are failing.

Agreed that for cluster-scoped resources, deletion of the Namespace should not have an effect. Is there some other aspect of Project deletion that we should be looking into?

dmage commented:
@sadasu I still don't see why you think it's related to project deletion. Project and Namespace aren't that different.

sadasu commented:
> @sadasu I still don't see why you think it's related to project deletion. Project and Namespace aren't that different.

Looking at the snippet of test logs attached in comment https://github.com/openshift/origin/pull/26566#issuecomment-965370718, I see that when the "oc explain" test fails, the Project has been deleted before "oc explain" for some of the resources have had a chance to run.

dmage commented:
@sadasu 

```
[BeforeEach] [sig-cli] oc explain
...
[It] should contain proper spec+status for CRDs [Suite:openshift/conformance/parallel]
...
Oct 28 23:41:15.350: INFO: Running 'oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1'
Oct 28 23:41:16.214: INFO: Error running /usr/bin/oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1:
...
[AfterEach] [sig-cli] oc explain
...
STEP: Destroying namespace "e2e-test-oc-explain-8d5ph" for this suite.
fail [github.com/openshift/origin/test/extended/cli/explain.go:450]: Unexpected error:
...
```

`[It] ...` marks that all `BeforeEach` blocks have successfully finished their job and the test has started. Then it runs `oc explain`, and it fails. Then Ginkgo runs `AfterEach` blocks, the last one deletes the namespace. And then it prints the error summary. It doesn't mean that the error happened after the namespace deletion. That's just how our test framework prints error details.

sadasu commented:
> @sadasu
> 
> ```
> [BeforeEach] [sig-cli] oc explain
> ...
> [It] should contain proper spec+status for CRDs [Suite:openshift/conformance/parallel]
> ...
> Oct 28 23:41:15.350: INFO: Running 'oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1'
> Oct 28 23:41:16.214: INFO: Error running /usr/bin/oc --namespace=e2e-test-oc-explain-8d5ph --kubeconfig=/tmp/configfile122745965 explain alertmanagerconfigs --api-version=monitoring.coreos.com/v1alpha1:
> ...
> [AfterEach] [sig-cli] oc explain
> ...
> STEP: Destroying namespace "e2e-test-oc-explain-8d5ph" for this suite.
> fail [github.com/openshift/origin/test/extended/cli/explain.go:450]: Unexpected error:
> ...
> ```
> 
> `[It] ...` marks that all `BeforeEach` blocks have successfully finished their job and the test has started. Then it runs `oc explain`, and it fails. Then Ginkgo runs `AfterEach` blocks, the last one deletes the namespace. And then it prints the error summary. It doesn't mean that the error happened after the namespace deletion. That's just how our test framework prints error details.

@dmage so if the order of these logs messages is not deterministic, how should we go about analyzing these failures? Cluster scoped resources cannot be "read" either when they are not present, or something to do with Project deletion or when the API server is unavailable. We know that the CRD is present. I think you are suggesting that Project deletion does not matter here. So, that leaves us with the API server being (temporarily) unavailable? Is there another explanation that I am missing?

dmage commented:
The order is deterministic, it's the test output followed by `fail [<file>:<line>]: <error details>`. You can ask `oc` maintainers for help.

sadasu commented:
After https://github.com/openshift/release/pull/23258 landed, "oc explain" tests failures are a rare occurrence. So, closing this.
